### PR TITLE
More programming content for TS/JS

### PR DESCRIPTION
**Pull request by rlisahuang**
I expect the build to fail for some examples because the way I constructed them does not work with Kevin's solution for TS inputs. Main problems:

- We cannot write top-level functions such as `solve()` and `prove()` in the [unit test](https://github.com/Z3Prover/z3/blob/master/src/api/js/src/high-level/high-level.test.ts) without bypassing type checking the function parameters
- `setParam` doesn't quite work
- We also would have to bypass type checking variable declarations if the types are within the `Z3` namespace.

@NikolajBjorner I should leave it to your judgement in terms of how to utilize/modify these examples so that everything works, or modify the `eval-z3.js` code as needed.

### CODE REVIEW

**Comment by bakkot** [comment]
> We cannot write top-level functions such as `solve()` and `prove()` in the [unit test](https://github.com/Z3Prover/z3/blob/master/src/api/js/src/high-level/high-level.test.ts) without bypassing type checking the function parameters

What goes wrong? I can probably suggest a solution.

> `setParam` doesn't quite work

That's because it exists on the global namespace, not on a specific context, which is where you're trying to invoke it. That is, you'll need to pull it off of the result of the `init` call, not off the `Context('main')` call. You can tweak [these lines](https://github.com/microsoft/z3guide/blob/18ccf2e63370bb7899f3065b527b06a3cb704d12/website/src/eval-z3/eval-z3.ts#L180-L188) as follows if you want to use it:

```js
  let wrapped = `
(function (Z3, setParam) { // note additional param
  'use strict';
  let module = {};
  ${result.result}
  return module.exports;
})
  `;
  return await (0, eval)(wrapped)(Z3.Context('main'), Z3.setParam); // note additional argument
```

and then invoke `setParam` directly (not as `Z3.setParam`).

> We also would have to bypass type checking variable declarations if the types are within the Z3 namespace.

You can add more type imports to the scope used for type checking by adding them to [this line](https://github.com/microsoft/z3guide/blob/18ccf2e63370bb7899f3065b527b06a3cb704d12/website/src/eval-z3/eval-z3.ts#L126). TypeScript is pretty good about type inference, though, so you should only rarely need to explicitly refer to a type (other than in function parameters, which it doesn't infer); you still get the benefits of type checking. Though you might want to explicitly refer to types just for documentation purposes.

---

Feel free to ping me about issues like this; I'm not necessarily going to be able to help but I'm happy to take a look.

File: website/docs-programming/02 - Z3 JavaScript/03 - more JS examples.md [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+const z = Z3.Real.const('z');
+
+const solver = new Z3.Solver();
+solver.add(x.mul(x).add(y.mul(y)).eq(1)); // x^2 + y^2 == 1
+solver.add(x.mul(x).mul(x).add(z.mul(z).mul(z)).lt('1/2')); // x^3 + z^3 < 1/2
+
+await solver.check(); // sat
+```
+
+TODO: seems that we can only get the final output, not intermediate ones
```
Comment (NikolajBjorner) [inline website/docs-programming/02 - Z3 JavaScript/03 - more JS examples.md:204]: console.log doesn't seem to work

File: website/docs-programming/02 - Z3 JavaScript/03 - more JS examples.md [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+// this solutions wraps around so we need to check using modulo
+(xv ^ yv) - 103n === (xv * yv) % 2n ** 32n; // true
+```
+
+```z3-js
+// AstVector
+const solver = new Z3.Solver();
+
+// @ts-ignore: skipping typechecking for the following line for now, is there a better solution?
```
Comment (NikolajBjorner) [inline website/docs-programming/02 - Z3 JavaScript/03 - more JS examples.md:255]: cast?

**Comment by rlisahuang** [comment]
Thanks for joining the conversation. 

> > We cannot write top-level functions such as `solve()` and `prove()` in the [unit test](https://github.com/Z3Prover/z3/blob/master/src/api/js/src/high-level/high-level.test.ts) without bypassing type checking the function parameters
> 
> What goes wrong? I can probably suggest a solution.
> 
The issue here is that type annotations are enforced unless you do `@ts-ignore`, and those types are not imported to the scope. I believe the issue can be addressed by your suggestion of type imports, though.

> That is, you'll need to pull it off of the result of the `init` call, not off the `Context('main')` call. You can tweak [these lines](https://github.com/microsoft/z3guide/blob/18ccf2e63370bb7899f3065b527b06a3cb704d12/website/src/eval-z3/eval-z3.ts#L180-L188) as follows if you want to use it:
> 
> ```js
>   let wrapped = `
> (function (Z3, setParam) { // note additional param
>   'use strict';
>   let module = {};
>   ${result.result}
>   return module.exports;
> })
>   `;
>   return await (0, eval)(wrapped)(Z3.Context('main'), Z3.setParam); // note additional argument
> ```
>
> and then invoke `setParam` directly (not as `Z3.setParam`).

The build would fail at `compileZ3JS`, so we don't even reach the evaluation. The complaint is that `setParam` does not exist on type `Context<'main'>`. I'm looking into the issue now, see if we can get around it without explicitly specifying the type of Z3.

File: website/src/eval-z3/eval-z3.ts [hunk 3]
```diff
@@ -122,12 +122,18 @@ function compile(
...
   let imports = `
-    import type { init as initT, Model, Solver } from 'z3-solver';
-    declare let init: typeof initT;
-    declare let { Context }: Awaited<ReturnType<typeof init>>;
-    declare let Z3: ReturnType<typeof Context<'main'>>;
-  `;
+  import type { init as initT, Model, Solver } from 'z3-solver';
+  declare let init: typeof initT;
+  declare let Z3: Awaited<ReturnType<typeof init>>;
+  declare let { Context, setParam } = Z3;
```
Comment (rlisahuang) [inline website/src/eval-z3/eval-z3.ts:135]: @bakkot I'm not sure I'm on the right track, so definitely would appreciate your help here. The example failed to be compiled is
```js
// non-linear arithmetic

Z3.setParam('pp.decimal', true);
Z3.setParam('pp.decimal_precision', 20);

const x = Z3.Real.const('x');
const y = Z3.Real.const('y');
const z = Z3.Real.const('z');

const solver = new Z3.Solver();
solver.add(x.mul(x).add(y.mul(y)).eq(1)); // x^2 + y^2 == 1
solver.add(x.mul(x).mul(x).add(z.mul(z).mul(z)).lt('1/2')); // x^3 + z^3 < 1/2

await solver.check(); // sat
```

Comment (bakkot) [reply website/src/eval-z3/eval-z3.ts]: Yeah, adding `declare let { Context, setParam } = Z3` should make the typechecker happy, and then including the change I suggested above should make the actual runtime behavior work.

Comment (rlisahuang) [reply website/src/eval-z3/eval-z3.ts]: Not really, I got
```
Error: main.ts (-2,39): Initializers are not allowed in ambient contexts.
main.ts (6,14): Property 'Real' does not exist on type '{ em: any; Z3: { mk_context: (c: Z3_config) => Z3_context; mk_context_rc: (c: Z3_config) => Z3_context; global_param_set: (param_id: string, param_value: string) => void; ... 719 more ...; qe_lite: (c: Z3_context, vars: Z3_ast_vector, body: Z3_ast) => Z3_ast; }; } & Z3HighLevel'.
main.ts (7,14): Property 'Real' does not exist on type '{ em: any; Z3: { mk_context: (c: Z3_config) => Z3_context; mk_context_rc: (c: Z3_config) => Z3_context; global_param_set: (param_id: string, param_value: string) => void; ... 719 more ...; qe_lite: (c: Z3_context, vars: Z3_ast_vector, body: Z3_ast) => Z3_ast; }; } & Z3HighLevel'.
main.ts (8,14): Property 'Real' does not exist on type '{ em: any; Z3: { mk_context: (c: Z3_config) => Z3_context; mk_context_rc: (c: Z3_config) => Z3_context; global_param_set: (param_id: string, param_value: string) => void; ... 719 more ...; qe_lite: (c: Z3_context, vars: Z3_ast_vector, body: Z3_ast) => Z3_ast; }; } & Z3HighLevel'.
main.ts (10,23): Property 'Solver' does not exist on type '{ em: any; Z3: { mk_context: (c: Z3_config) => Z3_context; mk_context_rc: (c: Z3_config) => Z3_context; global_param_set: (param_id: string, param_value: string) => void; ... 719 more ...; qe_lite: (c: Z3_context, vars: Z3_ast_vector, body: Z3_ast) => Z3_ast; }; } & Z3HighLevel'.
    at runMicrotasks (<anonymous>)
```

Comment (bakkot) [reply website/src/eval-z3/eval-z3.ts]: Oh, sorry, that's not legal syntax. Needs to be what it currently is, just with `setParam` also: `declare let { Context, setParam }: Awaited<ReturnType<typeof init>>;`

Comment (rlisahuang) [reply website/src/eval-z3/eval-z3.ts]: I messed up with the types in this commit. After further understanding the high-level API the issue is now fixed. The imports should be
```
  declare let { Context, setParam }: Awaited<ReturnType<typeof init>>;
  declare let Z3: ReturnType<typeof Context<'main'>>;
```
simply add `setParam` to the destructure of `init()`

**Comment by bakkot** [comment]
Also, if you want to make the `let model: Z3.Model` style of type names work, instead of explicitly importing all the types you need and referring to them with bare names like `Model`, I believe you can do that by adding

```ts
// @ts-ignore
import type * as Z3 from 'z3-solver';
```
to [the imports](https://github.com/microsoft/z3guide/blob/18ccf2e63370bb7899f3065b527b06a3cb704d12/website/src/eval-z3/eval-z3.ts#L125). (The `ts-ignore` is an unfortunate necessity due to https://github.com/microsoft/TypeScript/issues/48764#issuecomment-1102818406, but it doesn't actually disable the type checking in this instance.)

You'll need to also increment the index (currently 4) in [this line](https://github.com/microsoft/z3guide/blob/18ccf2e63370bb7899f3065b527b06a3cb704d12/website/src/eval-z3/eval-z3.ts#L147) so the rewriting still works.

**Comment by rlisahuang** [comment]
> Also, if you want to make the `Z3.Model` style of type names work, instead of explicitly importing all the types you need and referring to them with bare names like `Model`, I believe you can do that by adding
> 
> ```ts
> // @ts-ignore
> import type * as Z3 from 'z3-solver';
> ```

I'll let @NikolajBjorner decide which style of type names is desired in the examples.

**Comment by rlisahuang** [comment]
All problems in the extra examples fixed, ready to merge.

**Pull request merged**